### PR TITLE
[RFC] Add epoch_start_timestamp_ms to system object

### PR DESCRIPTION
**Pull request by andll**
This PR exposes timestamp of epoch start to the smart contracts in move, by putting this timestamp in the system object.

This can be useful for smart contract that needs some very coarse time before we have more granular timestamps.

### CODE REVIEW

**Comment by lxfind** [comment]
Might want to share in the move channel to see if this would be useful feature

File: crates/sui-framework/sources/governance/sui_system.move [hunk 1]
```diff
@@ -67,6 +67,8 @@ module sui::sui_system {
         validator_report_records: VecMap<address, VecSet<address>>,
         /// Schedule of stake subsidies given out each epoch.
         stake_subsidy: StakeSubsidy,
+        /// Unix timestamp of the current epoch start
+        epoch_start_timestamp_ms: u64,
```
Comment (amnn) [inline crates/sui-framework/sources/governance/sui_system.move:71]: This is probably worth exposing via an accessor (see L531 of this file for an example):

```rust
public fun epoch_start_timestamp_ms(self: &SuiSystemState): u64 {
    self.epoch_start_timestamp_ms
}
```

**Review by amnn — COMMENTED** [review]
This is definitely a useful feature, thanks for exposing this @andll!  It means that even folks using single-owner transactions have a way to get some concept of time -- we should make sure to document this as well, given how many folks have been asking for something like this, is that on your radar? cc @randall-Mysten 

I've left a comment about exposing the field with a getter function, so that folks can access it from within Move packages, but otherwise looks good to me!  Holding off on the accept as I'm less familiar with the epoch boundary code.

@andll, are you also planning on working on exposing checkpoint timestamps and/or generally narwhal consensus timestamps to Move, as part of the "Programmatic access to time" worksteram?

**Comment by andll** [comment]
> are you also planning on working on exposing checkpoint timestamps and/or generally narwhal consensus timestamps to Move, as part of the "Programmatic access to time" worksteram?

Yes, thats the plan :)

>I've left a comment about exposing the field with a getter function, so that folks can access it from within Move packages, but otherwise looks good to me! Holding off on the accept as I'm less familiar with the epoch boundary code.

Good point about accessor method, will add.
Do you know who would be a good person to review PR who might be more familiar with epoch boundary code? @emmazzz ?

**Comment by amnn** [comment]
>> are you also planning on working on exposing checkpoint timestamps and/or generally narwhal consensus timestamps to Move, as part of the "Programmatic access to time" worksteram?
>
> Yes, thats the plan :)

Awesome!  Okay, well for those, in case it's useful, to give you a sketch, it's sufficient to add it as an optional (i.e. `Option<u64>`) field of `TxContext`, which gets filled only for shared object transactions (I'm assuming that even if we are using checkpoint transactions, that we can't reliably pick a timestamp across validators for owned object transactions?).

There's a component that follows which involves adding a `Clock` type that forces a transaction to go to consensus etc, which I'm happy to take from there, but if you were also looking forward to doing that, I'm happy to sketch that bit out too!

> Do you know who would be a good person to review PR who might be more familiar with epoch boundary code?

I figured @lxfind, but I figured incorrectly, I'm happy to take a closer look!

**Comment by andll** [comment]
Added `epoch_start_timestamp_ms` and rebased

**Review by lxfind — COMMENTED** [review]
LGTM. I think it's worth adding this to the advance_epoch_safe_mode as well, given that this is unrelated to tokenomics.

**Comment by lxfind** [comment]
Actually no let's not bother..

**Review by lxfind — APPROVED** [review]

**Review by dariorussi — COMMENTED** [review]
I am just commenting for now and wanting to hear form @amnn as well... I will accept later otherwise.
This seems pretty good and easy for me. We are essentially moving some code earlier in a function in the checkpoint module, and pass the timestamp argument around so it reaches the system transaction created at epoch boundaries.
It assumes things will change at genesis time so I suppose that part is well understood and not a problem (not sure how we break up our releases "branches/forks").
I think we should be able to get that in no problem

**Review by dariorussi — APPROVED** [review]
well it has been approved so go for it 🙂

**Review by emmazzz — APPROVED** [review]
useful stuff!

**Pull request merged**